Re-land [Swift in WebKit] Work towards fixing the WTF clang module (part 2)

rr-codes opened the pull request:
#### 9e0290577c891d58e2649e15181c550c02afdfc1
<pre>
Re-land [Swift in WebKit] Work towards fixing the WTF clang module (part 2)
<a href="https://bugs.webkit.org/show_bug.cgi?id=304610">https://bugs.webkit.org/show_bug.cgi?id=304610</a>
<a href="https://rdar.apple.com/167045045">rdar://167045045</a>

Reviewed by Mike Wyrzykowski.

Fix the remaining issues in WTF, allowing for the module verifier to now be enabled.

* Source/WTF/Configurations/WTF.xcconfig:

- Fix the list of SDKs which actually support the WTF module
- Enable the library module verifier

* Source/WTF/Configurations/module.modulemap: Added.
* Source/WTF/WTF.xcodeproj/project.pbxproj:

Instead of auto-generating a module map, which itself was slightly ill-formed, just
write a normal module map by hand, using an umbrella directory.

* Source/WTF/wtf/InlineWeakRef.h:
* Source/WTF/wtf/RuntimeApplicationChecks.h:
* Source/WTF/wtf/StackTrace.cpp:
* Source/WTF/wtf/StackTrace.h:
* Source/WTF/wtf/SystemFree.h:
* Source/WTF/wtf/cocoa/AuditToken.h:
* Source/WTF/wtf/spi/cocoa/SecuritySPI.h:
* Source/WTF/wtf/spi/darwin/ProcessMemoryFootprint.h:
* Source/WTF/wtf/spi/darwin/ReasonSPI.h:

Add missing header includes and other module fixes

* Source/WebGPU/WebGPU/CommandEncoder.swift:
* Source/WebGPU/WebGPU/CxxBridging.h:
* Source/WebGPU/WebGPU/Internal/module.modulemap:

Workaround some issues in the current WebGPU module, which will soon be addressed once
the WebGPU module itself is fixed.

Canonical link: <a href="https://commits.webkit.org/304894@main">https://commits.webkit.org/304894@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a66d87b7b0c2f6ccf627623989d82363e54a1295

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136802 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9162 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/48089 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144533 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89775 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/723f4821-3bf4-4346-8f29-5a0c34a89446) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138674 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9864 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/9008 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104610 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89775 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/6ba74029-3c7d-4ac7-982b-3d26721738d8) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139747 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7208 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122568 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85447 "Found 1 new API test failure: WPE/TestWebKitWebContext:/webkit/WebKitWebContext/uri-scheme (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3e1e251e-4a7a-46cb-a302-ccf8bdcaba01) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6851 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4542 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5128 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/128763 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116172 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40755 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147291 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/135288 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8845 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41325 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/112974 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8863 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7432 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113297 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28780 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6773 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118855 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/63007 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8893 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36909 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/168068 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8614 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72459 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43848 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8833 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8685 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Source/WebKit/Configurations/WebKit.xcconfig:
```diff
@@ -389,27 +387,27 @@ EXCLUDED_MACOS_PLUGIN_FILE_NAMES[sdk=iphone*] = SecItemShim.dylib WebProcessShim
+WK_SUPPORTS_SWIFT_OBJCXX_INTEROP_watchos = ;
+WK_SUPPORTS_SWIFT_OBJCXX_INTEROP_watchsimulator = ;
+WK_SUPPORTS_SWIFT_OBJCXX_INTEROP_appletvos = ;
+WK_SUPPORTS_SWIFT_OBJCXX_INTEROP_appletvsimulator = ;
```
nit: you don't need to write these out, since they are undefined.

```suggestion
```

emw-apple commented on Source/WebKit/Configurations/WebKit.xcconfig:
```diff
@@ -389,27 +387,27 @@ EXCLUDED_MACOS_PLUGIN_FILE_NAMES[sdk=iphone*] = SecItemShim.dylib WebProcessShim
-WK_SUPPORTS_SWIFT_OBJCXX_INTEROP[sdk=macosx26.1*] = $(WK_NOT_$(DEPLOYMENT_LOCATION));
-WK_SUPPORTS_SWIFT_OBJCXX_INTEROP[sdk=macosx26.2*] = $(WK_NOT_$(DEPLOYMENT_LOCATION));
-WK_SUPPORTS_SWIFT_OBJCXX_INTEROP[sdk=macosx26.3*] = $(WK_NOT_$(DEPLOYMENT_LOCATION));
+WK_SUPPORTS_SWIFT_OBJCXX_INTEROP[sdk=macosx14*] = ;
+WK_SUPPORTS_SWIFT_OBJCXX_INTEROP[sdk=macosx15*] = ;
+WK_SUPPORTS_SWIFT_OBJCXX_INTEROP[sdk=macosx26.0*] = ;
+WK_SUPPORTS_SWIFT_OBJCXX_INTEROP[sdk=macosx26.1*] = ;
+WK_SUPPORTS_SWIFT_OBJCXX_INTEROP[sdk=macosx26.2*] = ;
+WK_SUPPORTS_SWIFT_OBJCXX_INTEROP[sdk=macosx26.3*] = ;
```
…but you _do_ need to write these out, so that they take precedence over the value in $(WK_SUPPORTS_SWIFT_OBJCXX_INTEROP_macosx).

emw-apple commented on Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
```diff
@@ -304,7 +304,11 @@ struct UIEdgeInsets;
 - (BOOL)_webView:(WKWebView *)webView gestureRecognizerCanBePreventedByTouchEvents:(UIGestureRecognizer *)gestureRecognizer WK_API_AVAILABLE(ios(16.5));
 
+#if TARGET_OS_WATCH
+- (void)_webView:(WKWebView *)webView startXRSessionWithFeatures:(_WKXRSessionFeatureFlags)features colorFormat:(NSUInteger)pixelFormat depthFormat:(NSUInteger)depthFormat completionHandler:(void (^)(id, UIViewController *))completionHandler WK_API_AVAILABLE(ios(17.4), visionos(1.1));
```
Could you comment more on this? It's not mentioned in the commit message at all.

Is `MTLPixelFormat` a typedef of NSUInteger?

rr-codes replied:
The Metal API is API on all platforms except watchOS, and so MTLPixelFormat is unavailable as API on watchOS, and so this is invalid in modules land in open source builds. And so I just replaced it with NSUInteger since it’s an NS_ENUM typedef’d as it

emw-apple commented on Source/WTF/Configurations/WTF.xcconfig:
```diff
@@ -47,29 +52,41 @@ SYSTEM_FRAMEWORK_SEARCH_PATHS = $(inherited) $(WK_PRIVATE_SDK_DIR)$(SYSTEM_LIBRA
+MODULE_VERIFIER_SUPPORTED_LANGUAGE_STANDARDS = c++2b;
 
-OTHER_MODULE_VERIFIER_FLAGS = -- -I$(BUILT_PRODUCTS_DIR)$(WK_LIBRARY_HEADERS_FOLDER_PATH) -iframework $(WK_PRIVATE_SDK_DIR)$(SYSTEM_LIBRARY_DIR)/PrivateFrameworks -Wsystem-headers-in-module=wtf;
+OTHER_MODULE_VERIFIER_FLAGS = -- -isystem $(DSTROOT)$(WTF_INSTALL_PATH_PREFIX)$(WK_LIBRARY_HEADERS_FOLDER_PATH) -I$(BUILT_PRODUCTS_DIR)$(WK_LIBRARY_HEADERS_FOLDER_PATH) -I$(WK_ADDITIONAL_SDKS)$(WK_LIBRARY_HEADERS_FOLDER_PATH) -isystem "$(WK_PRIVATE_SDK_DIR)$(WK_ALTERNATE_WEBKIT_SDK_PATH)$(WK_LIBRARY_HEADERS_FOLDER_PATH)" -iframework $(WK_PRIVATE_SDK_DIR)$(SYSTEM_LIBRARY_DIR)/PrivateFrameworks -Ws ...
```
Can you comment on what these extra paths are accomplishing? It worries me a little bit to see module verifier flags, since that implies that this module is actually _not_ safe for other clients to import, unless they have the same bespoke configuration that we do.

rr-codes replied:
some of them are to match the existing HEADER_SEARCH_PATHS and SYSTEM_FRAMEWORK_SEARCH_PATHS and all of the rest of those, and some are needed so that it finds the locally built WTF instead of the one in the SDK

emw-apple commented on Source/WTF/WTF.xcodeproj/project.pbxproj:
```diff
@@ -4350,7 +4352,7 @@
 			runOnlyForDeploymentPostprocessing = 0;
 			shellPath = /bin/sh;
-			shellScript = "if [ ${DEPLOYMENT_LOCATION} != YES ]; then\n    \"${SCRIPT_INPUT_FILE_0}\" --relative-to=\"${BUILT_PRODUCTS_DIR}${WTF_INSTALL_PATH_PREFIX}${WK_LIBRARY_HEADERS_FOLDER_PATH}\" \"${SCRIPT_INPUT_FILE_1}\"\nelse\n    \"${SCRIPT_INPUT_FILE_0}\" --relative-to=\"${DSTROOT}${PRIVATE_HEADERS_FOLDER_PATH}\"  \"${SCRIPT_INPUT_FILE_2}\"\nfi\n";
+			shellScript = "if [ ${DEPLOYMENT_LOCATION} != YES ]; then\n    \"${SCRIPT_INPUT_FILE_0}\" --relative-to=\"${BUILT_PRODUCTS_DIR}${WTF_INSTALL_PATH_PREFIX}${WK_LIBRARY_HEADERS_FOLDER_PATH}\" \"${SCRIPT_INPUT_FILE_1}\"\nelse\n    \"${SCRIPT_INPUT_FILE_0}\" --relative-to=\"<SDKROOT>${WK_LIBRARY_HEADERS_FOLDER_PATH}\"  \"${SCRIPT_INPUT_FILE_2}\"\nfi\n";
```
clever!

emw-apple reviewed (approved):
r=me (really r=ews and the internal testing you intend to complete). A couple questions inline, but not anything that needs to block forward progress.

rr-codes committed:
Re-land [Swift in WebKit] Work towards fixing the WTF clang module (part 2)
https://bugs.webkit.org/show_bug.cgi?id=304610
rdar://167045045

Reviewed by Mike Wyrzykowski.

Fix the remaining issues in WTF, allowing for the module verifier to now be enabled.

* Source/WTF/Configurations/WTF.xcconfig:

- Fix the list of SDKs which actually support the WTF module
- Enable the library module verifier

* Source/WTF/Configurations/module.modulemap: Added.
* Source/WTF/WTF.xcodeproj/project.pbxproj:

Instead of auto-generating a module map, which itself was slightly ill-formed, just
write a normal module map by hand, using an umbrella directory.

* Source/WTF/wtf/InlineWeakRef.h:
* Source/WTF/wtf/RuntimeApplicationChecks.h:
* Source/WTF/wtf/StackTrace.cpp:
* Source/WTF/wtf/StackTrace.h:
* Source/WTF/wtf/SystemFree.h:
* Source/WTF/wtf/cocoa/AuditToken.h:
* Source/WTF/wtf/spi/cocoa/SecuritySPI.h:
* Source/WTF/wtf/spi/darwin/ProcessMemoryFootprint.h:
* Source/WTF/wtf/spi/darwin/ReasonSPI.h:

Add missing header includes and other module fixes

* Source/WebGPU/WebGPU/CommandEncoder.swift:
* Source/WebGPU/WebGPU/CxxBridging.h:
* Source/WebGPU/WebGPU/Internal/module.modulemap:

Workaround some issues in the current WebGPU module, which will soon be addressed once
the WebGPU module itself is fixed.

Canonical link: https://commits.webkit.org/304894@main
